lib/client/db/postgres/repl/repl_test.go

```diff
@@ -0,0 +1,418 @@
+		"multiline with excessive whitespace": {lines: []string{"   SELECT", "    1", "     ;"}, expectedQuery: "SELECT\r\n1\r\n;"},
+		// Commands should only be executed if they are at the beginning of the
+		// first line.
+		"with command in the middle":              {lines: []string{"SELECT \\d 1;"}, expectedQuery: "SELECT \\d 1;"},
+		"multiline with command in the middle":    {lines: []string{"SELECT", "\\d", ";"}, expectedQuery: "SELECT\r\n\\d\r\n;"},
+		"multiline with command in the last line": {lines: []string{"SELECT", "1", "\\d;"}, expectedQuery: "SELECT\r\n1\r\n\\d;"},
+	} {
+		t.Run(name, func(t *testing.T) {
+			for _, line := range tt.lines {
+				writeLine(t, tc, line)
+			}
+
+			select {
+			case query := <-tc.QueryChan():
+				require.Equal(t, tt.expectedQuery, query)
+			case <-time.After(time.Second):
+				require.Fail(t, "expected to receive query but got nothing")
+			}
+
+			// Always expect a query reply from the server.
+			_ = readUntilNextLead(t, tc)
+		})
+	}
+}
+
+func writeLine(t *testing.T, c *testCtx, line string) {
+	t.Helper()
+	data := []byte(line + lineBreak)
+
+	// When writing to the connection, the terminal emulator always writes back.
+	// If we don't consume those bytes, it will block the ReadLine call (as
+	// we're net.Pipe).
+	go func(conn net.Conn) {
+		buf := make([]byte, len(data))
+		for {
+			n, err := conn.Read(buf[0:])
+			if err != nil {
+				t.Logf("Error while terminal reply on write: %s", err)
+				break
+			}
+
+			if string(buf[:n]) == line+lineBreak {
+				break
+			}
+		}
+	}(c.conn)
+
+	// Given that the test connections are piped a problem with the reader side
+	// would lead into blocking writing. To avoid this scenario we're using
+	// the Eventually just to ensure a timeout on writing into the connections.
+	require.EventuallyWithT(t, func(collect *assert.CollectT) {
+		_, err := c.conn.Write(data)
+		require.NoError(t, err)
+	}, 5*time.Second, time.Millisecond, "expected to write into the connection successfully")
+}
+
+// readUntilNextLead reads the contents from the client connection until we
+// reach the next leading prompt.
+func readUntilNextLead(t *testing.T, c *testCtx) string {
+	t.Helper()
+
+	var acc strings.Builder
+	for {
+		line := readLine(t, c)
+		if strings.HasPrefix(line, lineBreak+lineLeading(c.route)) {
+			break
+		}
+
+		acc.WriteString(line)
+	}
+	return acc.String()
+}
+
+func readLine(t *testing.T, c *testCtx) string {
+	t.Helper()
+
+	var n int
+	buf := make([]byte, 1024)
+	// Given that the test connections are piped a problem with the writer side
+	// would lead into blocking reading. To avoid this scenario we're using
+	// the Eventually just to ensure a timeout on reading from the connections.
+	require.EventuallyWithT(t, func(collect *assert.CollectT) {
+		var err error
+		n, err = c.conn.Read(buf[0:])
+		require.NoError(t, err)
+		require.Greater(t, n, 0)
```

Comment:
These need to be asserts, as this closure runs on a separate goroutine.

Also, should these assertions be made against `t` or against `collect`?